### DISCUSSION TITLE
iterator-based API not type-checker friendly

### DISCUSSION THREAD

**Discussion by zsol** [hynek/stamina#64] [General]
I was intrigued by the iterator + context manager based retry API offered by stamina (great library btw: love the docs, the overall APIs, the types-first approach 👍🏻), but after sprinkling it on [some code](https://github.com/zsol/py.wtf/commit/33668b63f04f743ddfcacc15812efa75498d0730) I was playing with, I noticed this pattern:

```py
foo = await fetch_foo()
print(foo)

# naturally (for me anyway) turns into

async for attempt in stamina.retry_context(Exception, 2):
    with attempt:
        foo = await fetch_foo()
print(foo)
```

But then type checkers rightfully complain on the last `print(foo)` call that `foo` might be unbound. After all, nothing in the types guarantee that the async generator yields anything.

So I have a few choices:

1. declare `foo` with some default value
2. move the `print` statement (any any further code) into the context manager scope
3. make a wrapper on top of `fetch_foo` specifically for this callsite and use the `stamina`'s decorator API

None of these are the end of the world, but it made me wonder if I'd ever use the iterator-based API again :) Has anyone else bumped into this? Is there a way to signal to a type checker that an iterator will definitely yield at least once?

**Comment by hynek** [comment]
I think I ran into it myself and was rather annoyed – but this is really a Python typing question: is there a way to tell the type checker that an iterator yields at least _one_ value? cc @Tinche 😇

**Reply by Tinche** [reply]
I think only pyright has a problem with this, Mypy seems to be OK with it.

I can't think of a way to make this work as-is. The `RetryingCaller` thing might be a better fit here?

**Reply by hynek** [reply]
Yeah, in this case

```python
foo = await stamina.RetryingCaller(attempts=2)(Exception, fetch_foo)
```

would be more idiomatic. But there's plenty of use-cases where the loop/contextmanager version is apt and the error isn't going away. :|

I'm tempted to lie to the type checker using something like `-> tuple[Attempt]`, but since it's an Iterator I wouldn't know how.

**Reply by hynek** [reply]
According to Łukasz, this only happens in Pyright’s strict mode? Unfortunately, it’s useless for real-world software IME. :|

**Reply by zsol** [reply]
Nah, this happens in Pyright's default mode as well as Pyre's. But yeah, mypy doesn't complain even in strict mode:
* [mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs&gist=eca5c4d37ca79c201399a49a9b4a2baa) (strict)
* [pyright](https://pyright-play.net/?code=JYWwDg9gTgLgBFAhgOwCYRAWAFA9QUwDM4RFhkAKASjgFoA%2BOAOQmXwC4c5u5Do4A%2BnHIIUAc3wUkaDADoAxgAsIweZIDaABgA0ARgC6VKp2w8zvCBDgBeOLq48o%2BAG75EAGwEwAnmEl8IKiA)
* [pyre](https://pyre-check.org/play?input=%23%20Pyre%20is%20being%20run%20in%20gradual%20typing%20mode%3A%20https%3A%2F%2Fpyre-check.org%2Fdocs%2Ftypes-in-python%2F%23gradual-typing%0A%23%20Use%20the%20%60%23%20pyre-strict%60%20header%20to%20run%20in%20strict%20mode%2C%20which%20requires%20annotations.%0Aimport%20random%0A%0Adef%20main()%20-%3E%20None%3A%0A%20%20%20%20for%20_%20in%20range(random.choice(%5B0%2C1%5D))%3A%0A%20%20%20%20%20%20%20%20foo%20%3D%201%0A%20%20%20%20reveal_type(foo)%0A%0A)

**Reply by ambv** [reply]
I can't reproduce in VScode with

```
    "python.analysis.typeCheckingMode": "basic",
```

I need to set it to "strict" for it to complain. This test file:

```py3
import stamina

async def g() -> None:
    async for attempt in stamina.retry_context(Exception, 2):
        with attempt:
            foo = 1
    print(foo + 1)
```

**Reply by zsol** [reply]
Ah you are right, `basic` doesn't include `reportPossiblyUnboundVariable`. But the pyright playground defaults to `standard` which does include that error (strict too). IDK, `basic` is a bit too ... basic IMO :) you don't get errors about incompatible overrides, for example. But that's not what this discussion is about

**Reply by hynek** [reply]
Yeah either way, I'm afraid this is a future typing PEP? :|
